Conversion from 'double' to 'float', possible loss of data errors

Hi,

Compiling with the last MSVS 2022, I got a lot _double to float, and possible loss from data_ errors.

I can see the project was used for VS 2010, that's way too undated and old.

_Changing Platform Toolset to "v143" (was "v100"): v100 = VS2010_

The errors:
5>C:\flac\src\libFLAC\format.c(580,110): warning C4334: '<<': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)
5>C:\flac\src\libFLAC\format.c(582,106): warning C4334: '<<': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)
5>C:\flac\src\libFLAC\format.c(584,74): warning C4334: '<<': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)
-----------
5>C:\flac\src\libFLAC\metadata_iterators.c(579,25): warning C4244: 'return': conversion from 'const __int64' to 'off_t', possible loss of data
5>C:\flac\src\libFLAC\metadata_iterators.c(1117,33): warning C4244: '+=': conversion from 'const __int64' to 'uint32_t', possible loss of data
5>C:\flac\src\libFLAC\metadata_iterators.c(1129,44): warning C4244: '=': conversion from '__int64' to 'uint32_t', possible loss of data
5>C:\flac\src\libFLAC\metadata_iterators.c(1152,35): warning C4244: '-=': conversion from 'const __int64' to 'uint32_t', possible loss of data
5>C:\flac\src\libFLAC\metadata_iterators.c(1647,34): warning C4244: '=': conversion from '__int64' to 'uint32_t', possible loss of data
5>C:\flac\src\libFLAC\metadata_iterators.c(1652,37): warning C4244: '=': conversion from '__int64' to 'uint32_t', possible loss of data
5>C:\flac\src\libFLAC\metadata_iterators.c(1666,36): warning C4244: '=': conversion from 'const __int64' to 'uint32_t', possible loss of data
-----------
5>C:\flac\src\libFLAC\metadata_object.c(1127,28): warning C4244: '=': conversion from 'FLAC__uint64' to 'uint32_t', possible loss of data
-----------
5>C:\flac\src\libFLAC\stream_encoder.c(3974,148): warning C4334: '<<': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)
5>C:\flac\src\libFLAC\stream_encoder.c(3976,145): warning C4334: '<<': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)
-----------
5>C:\flac\src\libFLAC\window.c(71,101): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(80,64): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(80,100): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(90,71): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(90,110): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(90,149): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(112,77): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(112,120): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(112,163): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(112,206): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(133,65): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(142,63): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(151,67): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(151,104): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(151,141): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(160,68): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(160,101): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(160,134): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(203,58): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(204,70): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(228,57): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(232,57): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(254,57): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(258,57): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(262,57): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
5>C:\flac\src\libFLAC\window.c(266,57): warning C4244: 'function': conversion from 'double' to 'float', possible loss of data
-----------
9>C:\flac\src\test_seeking\main.c(151,35): warning C4244: '=': conversion from '__int64' to 'uint32_t', possible loss of data
9>C:\flac\src\test_seeking\main.c(323,20): warning C4244: '=': conversion from '__int64' to 'long', possible loss of data
-----------
12>C:\flac\src\share\grabbag\picture.c(290,54): warning C4334: '<<': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)
12>C:\flac\src\share\grabbag\picture.c(308,59): warning C4244: 'function': conversion from 'const __int64' to 'FLAC__uint32', possible loss of data
-----------
20>C:\flac\src\test_streams\main.c(110,9): warning C4244: 'function': conversion from 'FLAC__uint64' to 'int', possible loss of data
20>C:\flac\src\test_streams\main.c(111,11): warning C4244: 'function': conversion from 'FLAC__uint64' to 'int', possible loss of data
20>C:\flac\src\test_streams\main.c(112,11): warning C4244: 'function': conversion from 'FLAC__uint64' to 'int', possible loss of data
20>C:\flac\src\test_streams\main.c(113,11): warning C4244: 'function': conversion from 'FLAC__uint64' to 'int', possible loss of data
========== Build: 26 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========

Please, consider to update to VS2022 now (yes, retail is out since March 31, 2022, and the last update is from this month).

--- a/src/libFLAC/format.c
+++ b/src/libFLAC/format.c
@@ -568,6 +568,11 @@ void FLAC__format_entropy_coding_method_partitioned_rice_contents_clear(FLAC__En
 	FLAC__format_entropy_coding_method_partitioned_rice_contents_init(object);
 }
 
+#if defined(_MSC_VER)
+// silence three MSVC warnings 'result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)'
+#pragma warning ( disable : 4334 )
+#endif
+
 FLAC__bool FLAC__format_entropy_coding_method_partitioned_rice_contents_ensure_size(FLAC__EntropyCodingMethod_PartitionedRiceContents *object, uint32_t max_partition_order)
 {
 	FLAC__ASSERT(0 != object);
@@ -585,3 +590,7 @@ FLAC__bool FLAC__format_entropy_coding_method_partitioned_rice_contents_ensure_s
 
 	return true;
 }
+
+#if defined(_MSC_VER)
+#pragma warning ( default : 4334 )
+#endif

--- a/src/libFLAC/metadata_iterators.c
+++ b/src/libFLAC/metadata_iterators.c
@@ -576,6 +576,10 @@ FLAC_API off_t FLAC__metadata_simple_iterator_get_block_offset(const FLAC__Metad
 	FLAC__ASSERT(0 != iterator);
 	FLAC__ASSERT(0 != iterator->file);
 
+#if defined(_MSC_VER)
+// silence MSVC warning 'conversion from 'const __int64' to 'off_t', possible loss of data'
+#pragma warning ( suppress : 4244 )
+#endif
 	return iterator->offset[iterator->depth];
 }
 
@@ -1099,6 +1103,11 @@ static FLAC__bool chain_merge_adjacent_padding_(FLAC__Metadata_Chain *chain, FLA
 		return false;
 }
 
+#if defined(_MSC_VER)
+// silence three MSVC warnings 'conversion from 'conversion from 'const __int64' to 'uint32_t', possible loss of data'
+#pragma warning ( disable : 4244 )
+#endif
+
 /* Returns the new length of the chain, or 0 if there was an error. */
 /* WATCHOUT: This can get called multiple times before a write, so
  * it should still work when this happens.
@@ -1175,6 +1184,10 @@ static FLAC__off_t chain_prepare_for_write_(FLAC__Metadata_Chain *chain, FLAC__b
 
 	return current_length;
 }
+
+#if defined(_MSC_VER)
+#pragma warning ( default : 4244 )
+#endif
 
 static FLAC__bool chain_read_cb_(FLAC__Metadata_Chain *chain, FLAC__IOHandle handle, FLAC__IOCallback_Read read_cb, FLAC__IOCallback_Seek seek_cb, FLAC__IOCallback_Tell tell_cb)
 {
@@ -1625,6 +1638,11 @@ typedef enum {
 	LBS_BLOCK_REMOVED
 } LastBlockState;
 
+#if defined(_MSC_VER)
+// silence three MSVC warnings 'conversion from 'conversion from 'const __int64' to 'uint32_t', possible loss of data'
+#pragma warning ( disable : 4244 )
+#endif
+
 FLAC_API FLAC__bool FLAC__metadata_chain_check_if_tempfile_needed(FLAC__Metadata_Chain *chain, FLAC__bool use_padding)
 {
 	/* This does all the same checks that are in chain_prepare_for_write_()
@@ -1701,6 +1719,10 @@ FLAC_API FLAC__bool FLAC__metadata_chain_check_if_tempfile_needed(FLAC__Metadata
 
 	return (current_length != chain->initial_length);
 }
+
+#if defined(_MSC_VER)
+#pragma warning ( default : 4244 )
+#endif
 
 FLAC_API FLAC__bool FLAC__metadata_chain_write(FLAC__Metadata_Chain *chain, FLAC__bool use_padding, FLAC__bool preserve_file_stats)
 {

--- a/src/libFLAC/metadata_iterators.c
+++ b/src/libFLAC/metadata_iterators.c
@@ -570,23 +570,14 @@ FLAC_API FLAC__bool FLAC__metadata_simple_iterator_is_last(const FLAC__Metadata_
 	return iterator->is_last;
 }
 
-#if defined(_MSC_VER)
-// silence MSVC warning 'conversion from 'const __int64' to 'off_t', possible loss of data'
-#pragma warning ( disable : 4244 )
-#endif
-
 /*@@@@add to tests*/
 FLAC_API off_t FLAC__metadata_simple_iterator_get_block_offset(const FLAC__Metadata_SimpleIterator *iterator)
 {
 	FLAC__ASSERT(0 != iterator);
 	FLAC__ASSERT(0 != iterator->file);
 
-	return iterator->offset[iterator->depth];
+	return (off_t)iterator->offset[iterator->depth];
 }
-
-#if defined(_MSC_VER)
-#pragma warning ( default : 4244 )
-#endif
 
 FLAC_API FLAC__MetadataType FLAC__metadata_simple_iterator_get_block_type(const FLAC__Metadata_SimpleIterator *iterator)
 {

--- a/src/libFLAC/metadata_iterators.c
+++ b/src/libFLAC/metadata_iterators.c
@@ -570,18 +570,23 @@ FLAC_API FLAC__bool FLAC__metadata_simple_iterator_is_last(const FLAC__Metadata_
 	return iterator->is_last;
 }
 
+#if defined(_MSC_VER)
+// silence MSVC warning 'conversion from 'const __int64' to 'off_t', possible loss of data'
+#pragma warning ( disable : 4244 )
+#endif
+
 /*@@@@add to tests*/
 FLAC_API off_t FLAC__metadata_simple_iterator_get_block_offset(const FLAC__Metadata_SimpleIterator *iterator)
 {
 	FLAC__ASSERT(0 != iterator);
 	FLAC__ASSERT(0 != iterator->file);
 
-#if defined(_MSC_VER)
-// silence MSVC warning 'conversion from 'const __int64' to 'off_t', possible loss of data'
-#pragma warning ( suppress : 4244 )
-#endif
 	return iterator->offset[iterator->depth];
 }
+
+#if defined(_MSC_VER)
+#pragma warning ( default : 4244 )
+#endif
 
 FLAC_API FLAC__MetadataType FLAC__metadata_simple_iterator_get_block_type(const FLAC__Metadata_SimpleIterator *iterator)
 {

--- a/src/libFLAC/metadata_object.c
+++ b/src/libFLAC/metadata_object.c
@@ -1103,6 +1103,11 @@ FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_append_spaced_point
 	return true;
 }
 
+#if defined(_MSC_VER)
+// silence MSVC warning 'conversion from 'FLAC__uint64' to 'uint32_t', possible loss of data'
+#pragma warning ( disable : 4244 )
+#endif
+
 FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_append_spaced_points_by_samples(FLAC__StreamMetadata *object, uint32_t samples, FLAC__uint64 total_samples)
 {
 	FLAC__ASSERT(object != NULL);
@@ -1124,10 +1129,6 @@ FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_append_spaced_point
 		if (num > 32768) {
 			/* Set the bound and recalculate samples accordingly. */
 			num = 32768;
-#if defined(_MSC_VER)
-// silence MSVC warning 'conversion from 'FLAC__uint64' to 'uint32_t', possible loss of data'
-#pragma warning ( suppress : 4244 )
-#endif
 			samples = total_samples / num;
 		}
 
@@ -1146,6 +1147,10 @@ FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_append_spaced_point
 
 	return true;
 }
+
+#if defined(_MSC_VER)
+#pragma warning ( default : 4244 )
+#endif
 
 FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_sort(FLAC__StreamMetadata *object, FLAC__bool compact)
 {

--- a/src/libFLAC/metadata_object.c
+++ b/src/libFLAC/metadata_object.c
@@ -1103,11 +1103,6 @@ FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_append_spaced_point
 	return true;
 }
 
-#if defined(_MSC_VER)
-// silence MSVC warning 'conversion from 'FLAC__uint64' to 'uint32_t', possible loss of data'
-#pragma warning ( disable : 4244 )
-#endif
-
 FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_append_spaced_points_by_samples(FLAC__StreamMetadata *object, uint32_t samples, FLAC__uint64 total_samples)
 {
 	FLAC__ASSERT(object != NULL);
@@ -1129,7 +1124,7 @@ FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_append_spaced_point
 		if (num > 32768) {
 			/* Set the bound and recalculate samples accordingly. */
 			num = 32768;
-			samples = total_samples / num;
+			samples = (uint32_t)(total_samples / num);
 		}
 
 		i = seek_table->num_points;
@@ -1147,10 +1142,6 @@ FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_append_spaced_point
 
 	return true;
 }
-
-#if defined(_MSC_VER)
-#pragma warning ( default : 4244 )
-#endif
 
 FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_sort(FLAC__StreamMetadata *object, FLAC__bool compact)
 {

--- a/src/libFLAC/metadata_object.c
+++ b/src/libFLAC/metadata_object.c
@@ -1124,6 +1124,10 @@ FLAC_API FLAC__bool FLAC__metadata_object_seektable_template_append_spaced_point
 		if (num > 32768) {
 			/* Set the bound and recalculate samples accordingly. */
 			num = 32768;
+#if defined(_MSC_VER)
+// silence MSVC warning 'conversion from 'FLAC__uint64' to 'uint32_t', possible loss of data'
+#pragma warning ( suppress : 4244 )
+#endif
 			samples = total_samples / num;
 		}
 

--- a/src/libFLAC/stream_encoder.c
+++ b/src/libFLAC/stream_encoder.c
@@ -3961,9 +3961,9 @@ uint32_t find_best_partition_order_(
 
 		/* save best parameters and raw_bits */
 		FLAC__format_entropy_coding_method_partitioned_rice_contents_ensure_size(prc, flac_max(6u, best_partition_order));
-		memcpy(prc->parameters, private_->partitioned_rice_contents_extra[best_parameters_index].parameters, sizeof(uint32_t)*(1<<(best_partition_order)));
+		memcpy(prc->parameters, private_->partitioned_rice_contents_extra[best_parameters_index].parameters, (uint32_t)sizeof(uint32_t)*(1<<(best_partition_order)));
 		if(do_escape_coding)
-			memcpy(prc->raw_bits, private_->partitioned_rice_contents_extra[best_parameters_index].raw_bits, sizeof(uint32_t)*(1<<(best_partition_order)));
+			memcpy(prc->raw_bits, private_->partitioned_rice_contents_extra[best_parameters_index].raw_bits, (uint32_t)sizeof(uint32_t)*(1<<(best_partition_order)));
 		/*
 		 * Now need to check if the type should be changed to
 		 * FLAC__ENTROPY_CODING_METHOD_PARTITIONED_RICE2 based on the

--- a/src/libFLAC/window.c
+++ b/src/libFLAC/window.c
@@ -42,6 +42,10 @@
 
 #ifndef FLAC__INTEGER_ONLY_LIBRARY
 
+#if defined(_MSC_VER)
+// silence 25 MSVC warnings 'conversion from 'double' to 'float', possible loss of data'
+#pragma warning ( disable : 4244 )
+#endif
 
 void FLAC__window_bartlett(FLAC__real *window, const FLAC__int32 L)
 {
@@ -278,5 +282,9 @@ void FLAC__window_welch(FLAC__real *window, const FLAC__int32 L)
 		window[n] = (FLAC__real)(1.0f - k * k);
 	}
 }
+
+#if defined(_MSC_VER)
+#pragma warning ( default : 4244 )
+#endif
 
 #endif /* !defined FLAC__INTEGER_ONLY_LIBRARY */

--- a/src/share/grabbag/picture.c
+++ b/src/share/grabbag/picture.c
@@ -287,7 +287,7 @@ static const char * read_file (const char * filepath, FLAC__StreamMetadata * obj
 	if (size < 0)
 		return error_messages[5];
 
-	if (size >= (1u << FLAC__STREAM_METADATA_LENGTH_LEN)) /* actual limit is less because of other fields in the PICTURE metadata block */
+	if (size >= (FLAC__off_t)(1u << FLAC__STREAM_METADATA_LENGTH_LEN)) /* actual limit is less because of other fields in the PICTURE metadata block */
 		return error_messages[11];
 
 	if ((buffer = safe_malloc_(size)) == NULL)
@@ -305,7 +305,7 @@ static const char * read_file (const char * filepath, FLAC__StreamMetadata * obj
 	}
 	fclose(file);
 
-	if (!FLAC__metadata_object_picture_set_data(obj, buffer, size, /*copy=*/false))
+	if (!FLAC__metadata_object_picture_set_data(obj, buffer, (FLAC__uint32)size, /*copy=*/false))
 		error_message = error_messages[6];
 	/* try to extract MIME type if user left it blank */
 	else if (*obj->data.picture.mime_type == '\0' && !local__extract_mime_type_(obj))

--- a/src/test_seeking/main.c
+++ b/src/test_seeking/main.c
@@ -148,7 +148,7 @@ static FLAC__bool read_pcm_(FLAC__int32 *pcm[], const char *rawfilename, const c
 		printf("ERROR: PCM verification requires 8 or 16 bps, got %u\n", bps);
 		return false;
 	}
-	samples = rawfilesize / channels / (bps>>3);
+	samples = (uint32_t)(rawfilesize / channels / (bps>>3));
 	if (samples > 10000000) {
 		fprintf(stderr, "ERROR: %s is too big\n", rawfilename);
 		return false;
@@ -320,7 +320,7 @@ static FLAC__bool seek_barrage(FLAC__bool is_ogg, const char *filename, FLAC__of
 	/* @@@ for is_ogg we should get it from last page's granulepos */
 	if(n == 0) {
 		/* 8 would imply no compression, 9 guarantees that we will get some samples off the end of the stream to test that case */
-		n = 9 * filesize / (decoder_client_data.channels * decoder_client_data.bits_per_sample);
+		n = (long int)(9 * filesize / (decoder_client_data.channels * decoder_client_data.bits_per_sample));
 	}
 
 	printf("Begin seek barrage, count=%u\n", count);

--- a/src/test_streams/main.c
+++ b/src/test_streams/main.c
@@ -104,6 +104,10 @@ static FLAC__bool write_little_endian_int32(FILE *f, FLAC__int32 x)
 }
 #endif
 
+#if defined(_MSC_VER)
+// silence 4 MSVC warnings 'conversion from 'FLAC__uint64' to 'int', possible loss of data'
+#pragma warning ( disable : 4244 )
+#endif
 static FLAC__bool write_little_endian_uint64(FILE *f, FLAC__uint64 x)
 {
 	return
@@ -117,6 +121,9 @@ static FLAC__bool write_little_endian_uint64(FILE *f, FLAC__uint64 x)
 		fputc(x >> 56, f) != EOF
 	;
 }
+#if defined(_MSC_VER)
+#pragma warning ( default : 4244 )
+#endif
 
 static FLAC__bool write_big_endian(FILE *f, FLAC__int32 x, size_t bytes)
 {

